docs(nx-cloud): add to launch templates documentation (#23156)
nrwl/nx: docs/nx-cloud/reference/launch-templates.md

# Launch Templates

A launch template defines the setup steps Nx Agents will run before running tasks. A custom launch template isn't required to use Nx Agents.
By default, Nx provides a simplified one for you. You can find built-in launch templates in the [`nx-cloud-workflows` repository](https://github.com/nrwl/nx-cloud-workflows/tree/main/launch-templates).

{% github-repository url="https://github.com/nrwl/nx-cloud-workflows/tree/main/launch-templates" title="Built-in Launch Templates" /%}

The easiest way to create a new custom launch template is to modify one of the built-in ones. To do that, create a file in the
`.nx/workflows` folder and copy one of the [built-in templates](https://github.com/nrwl/nx-cloud-workflows/blob/main/launch-templates/linux.yaml). You can name the file any way you want (e.g., `agents.yaml`) and customize the steps as needed.

This is an example of a launch template using all built-in features:

> ```yaml {% fileName="./nx/workflows/agents.yaml" %}

Author: I made this template way longer, which might be a lot at first, but wanted to provide a semi decent example file for people to build off with lots of comments.

launch-templates:
  # Custom template name, the name is referenced via --distribute-on="3 my-linux-medium-js"
  # You can define as many templates as you need, commonly used to make different sizes or toolchains depending on your workspace needs
  my-linux-medium-js:
    # see the available resource list below
    resourceClass: 'docker_linux_amd64/medium'
    # see the available image list below
    image: 'ubuntu22.04-node20.11-v6'
    # Define environment variables shared among all steps
    env:
      MY_ENV_VAR: shared
      # list out steps to run on the agent before accepting tasks
      # the agent will need a copy of the source code and dependencies installed
    init-steps:
      - name: Checkout
        # using a reusable step in an external GitHub repo,
        # this step is provided by Nx Cloud: https://github.com/nrwl/nx-cloud-workflows/tree/main/workflow-steps

>         uses: 'nrwl/nx-cloud-workflows/v3.6/workflow-steps/checkout/main.yaml'

Reviewer: Yes, but thats not ready yet. I can come around and update these after thats ready.

      - name: Restore Node Modules Cache
        uses: 'nrwl/nx-cloud-workflows/v3.6/workflow-steps/cache/main.yaml'
        # the cache step requires configuration via env vars
        # https://github.com/nrwl/nx-cloud-workflows/tree/main/workflow-steps/cache#options
        env:
          KEY: 'package-lock.json|yarn.lock|pnpm-lock.yaml'
          PATHS: 'node_modules'

>           BASE_BRANCH: 'main'

Reviewer: Yes, we should be able to. But the examples in the repo are not updated, and I'm writing the documentation for this still.

```
nx-cloud start-ci-run --distribute-on="8 linux-medium-js" --with-env-vars="VAR1,VAR2"
```

Or pass all the environment variables except OS-specific ones with this `--with-env-vars="auto"`:

```
nx-cloud start-ci-run --distribute-on="8 linux-medium-js" --with-env-vars="auto"
```

## Pass Values Between Steps

If you need to pass a value from one step to another step such as assigning the value to an existing or new environment variable. You can write to the `NX_CLOUD_ENV` environment file.

Commonly this is used for redefining the `PATH` or setting options for various toolchains.

```yaml {% fileName="./nx/workflows/agents.yaml" %}
launch-templates:
  my-template-name:
    init-steps:
      - name: Set PATH
        script: echo "PATH=$HOME/.cargo/bin:$PATH" >> $NX_CLOUD_ENV
      - name: Check PATH
        script: |
          # now contains $HOME/.cargo/bin
          echo $PATH 
          # can invoke cargo directly because it's in the PATH now. 
          cargo --version
```

> ## Private NPM Registry

Author: still need to confirm this works.

Author: works

```yaml {% fileName=".nx/workflows/agents.yaml" %}
launch-templates:
  node-18:
    resourceClass: 'docker_linux_amd64/medium'
    image: 'ubuntu22.04-node20.11-v6'
    init-steps:
      - name: Checkout
        uses: 'nrwl/nx-cloud-workflows/v3.6/workflow-steps/checkout/main.yaml'
      - name: Install nvm
        script: |
          # run nvm install script
          curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.7/install.sh | bash
          # source the updated profile to get the nvm command available
          source ~/.profile
          # install the needed version of node via nvm
          nvm install 21.7.3
          # echo the current path (which now includes the nvm-provided version of node) to NX_CLOUD_ENV
          echo "PATH=$PATH" >> $NX_CLOUD_ENV
      - name: Print node version
        # confirm that the node version has changed
        script: node -v
      - name: Install Node Modules
        uses: 'nrwl/nx-cloud-workflows/v3.6/workflow-steps/install-node-modules/main.yaml'
      # Continue setup steps as needed

> ```

Reviewer 2: oh nice! I think the plan is to eventually bake nvm into the image, and allow users to chose their version via a custom step, similar to [how github actions do it](https://github.com/actions/setup-node?tab=readme-ov-file#usage) 

but this will be perfect to have until then!

Author: I can take a stab at it monday.

## Installing various packages

You can use `apt` to install popular linux packages. This is helpful in streamlining setting up various toolchains needed for your workspace.

For example, you can install the [GitHub CLI](https://cli.github.com/) on the agents if needed.

```yaml {% fileName="./nx/workflows/agents.yaml" %}
launch-templates:
  my-linux-medium-js:
    resourceClass: 'docker_linux_amd64/medium'
    image: 'ubuntu22.04-node20.11-v6'
    init-steps:
      - name: Install GH CLI
        script: |
          sudo apt install gh -y
```

{% callout type="note" title="Installing without apt" %}
If you're trying to install a package that isn't available on `apt`, check that packages install steps for Debian base linux. Usually there are a handful of installation scripts that can be used [similar to `nvm`](#custom-node-version)
{% /callout %}

## Dynamic Changesets

> Nx Agents support dynamically distributing a given changeset for a CI run with a different number of agents and templates.
> This is equivalent of changing the value passed to `--distribute-on` based on your changes. Instead of defining a static number of agents for each change.
> 
> Currently, you can define a `small`, `medium` and `large` changeset for Nx Agents to use.
> These sizes are based on the ratio of affected projects Nx Cloud calculates for your PR.

Reviewer 2: ```suggestion
NxCloud can calculate how big your pull request is based on how many projects in your workspace it affects. You can then configure Nx Agents to dynamically use a different number of agents based on your changeset size.

Here we define a `small`, `medium` and `large` distribution strategy:
```

Reviewer 2: you don't need to use my suggestion, was just trying to see if we could shorten the explanation

```yaml {% fileName="./nx/workflows/agents.yaml" %}
launch-templates:
  my-template-name:
    init-steps:
      - name: Set PATH
        script: echo "PATH=$HOME/.cargo/bin:$PATH" >> $NX_CLOUD_ENV
      - name: Check PATH
        script: |
          # now contains $HOME/.cargo/bin
          echo $PATH 
          # can invoke cargo directly because it's in the PATH now. 
          cargo --version
```

## Private NPM Registry

If your project consumes packages from a private registry, you'll have to set up an authentication step in a custom launch template. We recommend creating an `.npmrc` file. You can pass the auth token from your main agent, so it's available to the agent machines.

```yaml {% fileName=".nx/workflows/agents.yaml" %}
launch-templates:
  my-linux-medium-js:
    resourceClass: 'docker_linux_amd64/medium'
    image: 'ubuntu22.04-node20.11-v7'
    init-steps:
      - name: Checkout
        uses: 'nrwl/nx-cloud-workflows/v3.6/workflow-steps/checkout/main.yaml'
      - name: Auth to Registry
        script: |
          echo "@myorg:registry=https://my-private-registry.com/myorg" >> .npmrc

>           echo "//my-private-registry.com/:_authToken=MY_REGISTRY_TOKEN" >> .npmrc

Reviewer 2: it would be cool if we could create a custom step to handle this 🤔

Author: I agree, though might be a bit more verbose to define inputs for a step than just a few script lines. 

Was thinking if we could take the .npmrc or .yarnrc file if present on the main agent and sync it to the agents so people only configure it via the main agent and wouldn't even need to use a custom launch template. but that edges on the line of "too much magic" for me IMO. 

At min I think a documentation is at min a good step forward (I still need to confirm this works as expected)

# Launch Templates

A launch template defines the setup steps Nx Agents will run before running tasks. A custom launch template isn't required to use Nx Agents.

> By default, Nx provides a simplified one for you. You can find built-in launch templates in the [`nx-cloud-workflows` repository](https://github.com/nrwl/nx-cloud-workflows/tree/main/launch-templates).

Reviewer 3: ```suggestion
Nx Cloud provides several pre-built launch templates for common use-cases. You can view available templates in the [`nx-cloud-workflows` repository](https://github.com/nrwl/nx-cloud-workflows/tree/main/launch-templates).
```

> {% github-repository url="https://github.com/nrwl/nx-cloud-workflows/tree/main/launch-templates" title="Built-in Launch Templates" /%}

Reviewer 3: if we think `pre-built` is better then we can update the title here too

## Getting Started

The easiest way to create a new custom launch template is to modify one of the built-in ones. To do that, create a file in the
`.nx/workflows` folder and copy one of the [built-in templates](https://github.com/nrwl/nx-cloud-workflows/blob/main/launch-templates/linux.yaml). You can name the file any way you want (e.g., `agents.yaml`) and customize the steps as needed.

## Launch Template Structure

### `launch-templates`

A `map` of launch template configurations. This value is required.

```yaml {% fileName=".nx/workflows/agents.yaml" %}
launch-templates:
```

### `launch-templates.<template-name>`

> Name of your custom launch template. This name is used via `--distribute-on="<# of agents> <template-name>"` when starting the ci run. Multiple launch templates are not required. You can define as many as you need.

Reviewer 3: ```suggestion
The name of your custom launch template. This name is used via `--distribute-on="<# of agents> <template-name>"` when starting the ci run. Supports one to many uniquely-named launch templates.
```

Multiple launch templates can be useful for setting up different toolchains (rust, java, node versions) or resources classes for your workspace needs.

```yaml {% fileName=".nx/workflows/agents.yaml" %}
launch-templates:
  template-one:
  template-two:
```

```
nx-cloud start-ci-run --distribute-on="3 template-one"
```

### `launch-templates.<template-name>.resource-class`

> The `launch-templates.<template-name>.resource-class` defines the memory and vCPUs available to the agent machine.

Reviewer 3: ```suggestion
A launch template's `resource-class` defines the memory and vCPUs available to each agent machine.
```

- `docker_linux_arm64/extra_large`
- `windows/medium`

See their detailed description and pricing at [nx.app/pricing](https://nx.app/pricing#plan-detail?sutm_source=nx.dev&utm_medium=launch-templates).

### `launch-templates.<template-name>.image`

> The `launch-templates.<template-name>.image` defines the available base software for the agent machine.

Reviewer 3: ```suggestion
A launch template's `image` defines the available base software for the agent machine.
```

```yaml {% fileName=".nx/workflows/agents.yaml" %}
launch-templates:
  template-one:
    image: 'ubuntu22.04-node20.11-v7'
```

> Nx Cloud provides the following images:

Reviewer 3: I don't know that we want to list every image here. Most of the updates we have published are out of necessity, so we probably don't want someone to pick an arbitrary older version for whatever reason.

We might just want to list the most up-to-date version and revisit the changelog/feature lists as a separate item.

Author: I'll just list from v5 onwards since that's the ones we actually have "notes" for. a more verbose changelog/release notes will be a nice enhancement down the road.

> Changes added in previous images are included in newer images unless otherwise denoted

- `ubuntu22.04-node20.9-v1`
- `ubuntu22.04-node20.9-v2`
- `ubuntu22.04-node20.9-v3`
- `ubuntu22.04-node20.11-v1`
- `ubuntu22.04-node20.11-v2`
- `ubuntu22.04-node20.11-v3`
- `ubuntu22.04-node20.11-v4`
- `ubuntu22.04-node20.11-v5`
  - added elevated permission access via `sudo`
- `ubuntu22.04-node20.11-v6`
  - `corepack` is enabled by default
- `ubuntu22.04-node20.11-v7`
  - added java version 17
- `windows-2022`

> Note: Windows-based images can only run on Windows-based [resource classes](#launch-templatestemplate-nameresourceclass).

Enterprise accounts can use custom images.

### `launch-templates.<template-name>.env`

> A `map` of values to be available within **all** steps of the specific launch template.

Reviewer 3: ```suggestion
A launch template's `env` defines a `map` of environment variable names and values to be available within **all** steps of the specific launch template.
```

```yaml {% fileName=".nx/workflows/agents.yaml" %}
launch-templates:
  template-one:
    env:
      MY_ENV_VAR: 'my-var-value'
```

### `launch-templates.<template-name>.init-steps`

> A launch template defines a series of steps to set up an agent. Without a defined `init-steps` the Nx Agent is unable to process any tasks. This includes things such as checking out your workspace source code and installing any necessary dependencies. Any extra setup your workspace needs to run should be defined as a step. Once all steps run, the agent machine will inform Nx Cloud that it is ready to accept tasks.

Reviewer 3: ```suggestion
A launch template's `init-steps` defines the series of steps to perform before an Nx Agent runs. Without a defined `init-steps` the Nx Agent is unable to process any tasks. 

Typical `init-steps` perform actions such as checking out your workspace source code and installing any necessary dependencies. Any extra setup your workspace needs to run should be defined as an `init-step`. Once all steps run successfully, the agent machine will inform Nx Cloud that it is ready to accept tasks.
```

```yaml {% fileName=".nx/workflows/agents.yaml" %}
launch-templates:
  template-one:
    init-steps:
```

### `launch-templates.<template-name>.init-steps[*].name`

> Name of the step, this will be reflected in the Nx Cloud UI. `name` can be used in conjunction with [`uses`](#launch-templatestemplate-nameinit-stepsuses) and [`script`](#launch-templatestemplate-nameinit-stepsscript)

Reviewer 3: ```suggestion
An init-step's `name` is the label that will be reflected in the Nx Cloud UI. `name` can be used in conjunction with [`uses`](#launch-templatestemplate-nameinit-stepsuses) and [`script`](#launch-templatestemplate-nameinit-stepsscript)
```

```yaml {% fileName=".nx/workflows/agents.yaml" %}
launch-templates:
  template-one:
    init-steps:
      - name: 'My Helpful Step Name'
```

### `launch-templates.<template-name>.init-steps[*].uses`

> Use a predefined step file in your `init-steps`.

Reviewer 3: ```suggestion
When defined, specifies an existing step file in your to be used. **Cannot be used when `script` is also defined**
```

You can find the [list of Nx Cloud reusable steps here](https://github.com/nrwl/nx-cloud-workflows/tree/main/workflow-steps).

```yaml {% fileName=".nx/workflows/agents.yaml" %}
launch-templates:
  template-one:
    init-steps:
      - uses: 'nrwl/nx-cloud-workflows/v3.6/workflow-steps/checkout/main.yaml'
      - name: 'Install Node Modules'
        uses: 'nrwl/nx-cloud-workflows/v3.6/workflow-steps/install-node-modules/main.yaml'
```

### `launch-templates.<template-name>.init-steps[*].script`

> Run an inline script using the default shell of the agent.

Reviewer 3: ```suggestion
When defined, allows an inline script to be run. **Cannot be used when `uses` is also defined**
```

```yaml {% fileName=".nx/workflows/agents.yaml" %}
launch-templates:
  template-one:
    init-steps:
      - name: 'Print Node Version and PATH'
        script: |
          node -v
          echo $PATH
```

### `launch-templates.<template-name>.init-steps[*].env`

> Similar to the [`launch-template.<template-name>.env`](#launch-templatestemplate-nameenv), except this environment variable `map` only is set for the current step.

Reviewer 3: ```suggestion
An init-step's `env` is similar to the [`launch-template.<template-name>.env`](#launch-templatestemplate-nameenv), except the environment variable `map` only scoped for the current step only.
```

- `ubuntu22.04-node20.9-withDind-v2`
- `ubuntu22.04-node20.9-v3`
- `ubuntu22.04-node20.9-withDind-v3`
- `ubuntu22.04-node20.9-withDind-v4`
- `ubuntu22.04-node20.9-withDind-v5`
- `ubuntu22.04-node20.9-withDind-v6`
- `ubuntu22.04-node20.9-withDind-v7`

> - `ubuntu22.04-node20.9-withDind-v8`

Author: what happened with dind and node versions? are these out of sync?

Reviewer 2: I would not talk about dind yet - it currently only works for Enterprises, and we can point them to it if they need it.

And it might confuse Pro users trying to use these images and wondering why Docker doesn't work.

I have an item this cycle to look at enabling this for all Pro users, so we can revisit then?

Author: removed dind from image list

- `ubuntu22.04-node20.11-v1`
- `ubuntu22.04-node20.11-v2`
- `ubuntu22.04-node20.11-v3`
- `ubuntu22.04-node20.11-v4`
- `ubuntu22.04-node20.11-v5`
  - introduced elevated permission access via `sudo`
- `ubuntu22.04-node20.11-v6`

>   - `corepack` is enabled by default

Author: do we have more release notes worth bring up?
